### PR TITLE
collectd exec script to show process-specific metrics

### PR DESCRIPTION
**Pull request by deepix**
- requires psutil
- sends metrics in PUTVAL format
- also sends metrics as a notification (PUTNOTIF)
- sends metrics only for some processes (memory and cpu thresholds)

### CODE REVIEW

**Comment by tedoc2000** [comment]
I think this should be a python plugin.. instead of an external script to execute.
You can find info on the python plugin here:
https://collectd.org/documentation/manpages/collectd-python.5.shtml

**Comment by mpetazzoni** [comment]
I agree with @tedoc2000, this should be a Python plugin. You can check out https://github.com/signalfx/docker-collectd-plugin for an example.

Also make sure your branch history is clean; we don't want to see duplicate commits/fix up commits/etc. Rebase/clean-up before you push your branch.

**Comment by deepix** [comment]
It's now a python plugin.  There was a mess because I'd forked the repo - thanks to Max for cleaning it all up.  Please review and let me know your comments.  I'm new to python so would appreciate if there are easier language-specific constructs or idioms that I should be using.

File: scripts/collectd-process-watch/collectd-process-watch.db [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+process.cpu.percent	value:GAUGE:0:100
+process.bytes.read	value:COUNTER:U:U
+process.runtime.seconds	value:COUNTER:U:U
+container.bytes.write	value:COUNTER:U:U
+container.runtime.seconds	value:COUNTER:U:U
+container.network.bytes.tx	value:COUNTER:U:U
+container.mem.percent	value:GAUGE:0:100
+container.cpu.percent	value:GAUGE:0:100
+container.bytes.read	value:COUNTER:U:U
+container.network.bytes.rx	value:COUNTER:U:U
```
Comment (wt) [inline scripts/collectd-process-watch/collectd-process-watch.db]: Modifying the types db is not really that great. If you can avoid it, you should. For example, you should just use existing generic types, if possible. For example, just use "counter" for counters.

**Comment by mpetazzoni** [comment]
@deepix Thanks for your work on this.

* Is the point of collecting container metrics to allow people who use containers without Docker to still get metrics? Is that really a significant population of user to justify this?
* For the process metrics, are we capturing here anything that the `processes` plugin doesn't already get?

If the answer to both questions is yes, then we can continue the work on this. But in this case, as discussed, this needs to be in its own repository. Create a new (private for now) repository under the `signalfx` organization, and call it `collectd-ps-plugin`. Initialize it with a dummy README file, then move your code over there into a branch and submit a PR on that repo. We'll move the discussion there when that's done.

Thanks!

**Comment by deepix** [comment]
@mpetazzoni There are couple drawbacks with the processes plugin:

(1) it takes a config-time regex pattern to filter processes, but ours filters on thresholds of CPU and memory.
(2) it does not have awareness of containers.

Container support was requested by Phil in order to give a more complete snapshot of the system.  The plugin only supports Docker - it essentially stitches /proc and docker stats (via its Unix socket /var/run/docker.sock) together to tell which process belongs to which container.  (Plain ps only shows cgroup.)

**Comment by wt** [comment]
Could (2) be added to the collectd processes plugin?

**Comment by deepix** [comment]
It's clumsy and non-standard.  For Docker, you have to read the cgroup file, extract container ID, connect to the docker socket and issue a REST API call for "docker inspect" for the extracted ID.  The networking also adds some delay.  I don't know if the processes plugin will take this change in its current state - maybe they will wait a bit for it to stabilize and standardize.

**Comment by mpetazzoni** [comment]
It doesn't seem like your plugin correlates Docker metrics with process metrics. Which means you've re-implemented the `docker-collectd-plugin`, with all the mistakes it made in its beginnings (for example, you didn't take into account that calling `dclnt.stats()` can take up to 1 second, so your plugin isn't going to work well when you have more than a few containers).

My point is the part about collecting Docker container metrics is already covered by the `docker-collectd-plugin`. When you take that out, the rest is just about the processes, something the `processes` plugin does relatively well already. I understand we do filtering differently, and maybe that's interesting, but I don't think it warrants a distinct plugin vs contributing this upstream to the `processes` plugin.

**Pull request closed without merging**